### ISSUE TITLE
[Github] Add support for Github App authentication

### ISSUE THREAD

**Issue by jamiehynds** [elastic/integrations#7351]
Our Github integration relies on a Personal Access Token (PAT) to authenticate against their API, however Github recommends using the Github app for authentication against their API: 

_If you want to use the API on behalf of an organization or another user, GitHub recommends that you use a GitHub App. For more information, see "[About authentication with a GitHub App](https://docs.github.com/en/apps/creating-github-apps/authenticating-with-a-github-app/about-authentication-with-a-github-app)."

You can also create an OAuth token with an OAuth app to access the REST API. However, GitHub recommends that you use a GitHub App instead. GitHub Apps allow more control over the access and permission that the app has._

https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/about-authentication-to-github#authenticating-to-the-api-with-an-app

This issue will track our progress towards supporting Github App authentication.

**Comment by chrisberkhout** [comment]
### Background

The general flow of Github App auth is:

* Use a configured private key to generate a signed JSON Web Token (JWT) locally
* Within 10 mins use the JWT to fetch an Installation Access Token (IAT)
* For the next 1 hour use the IAT to do work, then repeat

One way to do this is to have CEL generate the JWT and fetch the IAT as required, but for that it needs to support RS256 signing (Asymmetric RSA + SHA-256), which it currently does not (see existing [Crypto functions](https://pkg.go.dev/github.com/elastic/mito/lib#Crypto)). In Go, the signing would use [`rsa.SignPKCS1v15`](https://pkg.go.dev/crypto/rsa#SignPKCS1v15). Alternatively, it could be done with a higher-level JWT library. 

The Github integration uses CEL for the `security_advisories` data stream, but for the others it uses HTTP JSON: `audit`, `code_scanning`, `dependabot`, `issues`, `secret_scanning`.

The data streams `code_scanning`, `dependabot`, `issues`, `secret_scanning` do not use a persisted cursor. The `audit` data stream does use a persisted cursor and can fetch an unlimited amount of history (default: 730 hours) but it does fingerprinting for de-duplication. I think these are all good candidates for migration to CEL.

### Plan

* Add the required functionality to CEL
* Migrate the Github integration's 5 remaining HTTP JSON data streams to CEL and add Github App auth support

### Alternatives

I think these alternatives are less good than the plan:

* Add RS256 / JWT signing support to HTTP JSON in templating functions
* Add Github App auth as a background auth mechanism in HTTP JSON
* Add Github App auth as a background auth mechanism in CEL (and still migrate data streams)

**Comment by chrisberkhout** [comment]
Follow CEL migration strategy used in https://github.com/elastic/integrations/pull/14787

**Comment by chrisberkhout** [comment]
### What about the migration to CEL  makes it a breaking change?

The data streams to migrate have settings at the data stream level that will need to be re-entered to continue operation when upgrading to the migrated package.

The `audit` data stream also has a persisted cursor that will be reset, but with a breaking change notification we can let existing users know that setting a shorter Initial Interval will avoid re-fetching a lot of history (although if it is fetched, the fingerprinting should avoid duplication).

We can consolidate the API-related inputs into one (currently split between "Collect GitHub logs via API" and "Collect GitHub Security Advisories data via API") and make the credential settings be shared between all CEL data streams so they don't have to be re-entered separately for each data stream.

### What will the user experience be?

When upgrading, the user will see a breaking change warning. The HTTP JSON input stream will be gone and the CEL input stream will be present and disabled. They will need to re-enable the input stream and re-enter credentials to keep getting data. Duplicate data won't be a problem.

### Decision

@cpascale43 is this an acceptable user impact for the benefits of consolidating inputs and unblocking Github App authentication support?

**Comment by cpascale43** [comment]
Yes @chrisberkhout, exactly - I think the goal is to avoid the fragmented experience described here: https://github.com/elastic/integrations/issues/14813

Does the breaking change affect just the `audit` data stream or also `code_scanning`, `dependabot`, `issues`, `secret_scanning`? Does it introduce the potential for any gaps in coverage? If we do proceed I'd just want to give users advance notice through community slack, discussion forum etc so they can plan ahead

**Comment by chrisberkhout** [comment]
@cpascale43 

> Does the breaking change affect just the `audit` data stream or also `code_scanning`, `dependabot`, `issues`, `secret_scanning`? Does it introduce the potential for any gaps in coverage? If we do proceed I'd just want to give users advance notice through community slack, discussion forum etc so they can plan ahead

They will all need user action to continue working. There will be the breaking change notice in the UI, but we can time the release to allow notification by other channels as well.

Data gaps shouldn't be a issue if it's reactivated during the upgrade.
